From bdda54cc39843589ee91a0176ca9a94adf307763 Mon Sep 17 00:00:00 2001
From: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Date: Sat, 20 Feb 2021 11:01:02 +0800
Subject: [PATCH] kprobes/arm64: Blacklist sdei watchdog callback functions
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: bdda54cc39843589ee91a0176ca9a94adf307763


hulk inclusion
category: feature
bugzilla: 48046
CVE: NA

-------------------------------------------------------------------------

Functions called in sdei_handler are not allowed to be kprobed, so
marked them as NOKPROBE_SYMBOL. There are so many functions in
'watchdog_check_timestamp()'. Luckily, we don't need
'CONFIG_HARDLOCKUP_CHECK_TIMESTAMP' now. So just make
CONFIG_SDEI_WATCHDOG depends on !CONFIG_HARDLOCKUP_CHECK_TIMESTAMP
in case someone add 'CONFIG_HARDLOCKUP_CHECK_TIMESTAMP' in the future.

Signed-off-by: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Reviewed-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Xiongfeng Wang <wangxiongfeng2@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Geliang Tang <geliang.tang@suse.com>
---
 arch/arm64/kernel/watchdog_sdei.c | 2 ++
 kernel/watchdog.c                 | 2 ++
 kernel/watchdog_hld.c             | 2 ++
 lib/Kconfig.debug                 | 2 +-
 4 files changed, 7 insertions(+), 1 deletion(-)

--- a/arch/arm64/kernel/watchdog_sdei.c
+++ b/arch/arm64/kernel/watchdog_sdei.c
@@ -14,6 +14,7 @@
 #include <asm/sdei.h>
 #include <asm/virt.h>
 #include <linux/arm_sdei.h>
+#include <linux/kprobes.h>
 #include <linux/nmi.h>
 
 /* We use the secure physical timer as SDEI NMI watchdog timer */
@@ -66,6 +67,7 @@ static int sdei_watchdog_callback(u32 ev
 
 	return 0;
 }
+NOKPROBE_SYMBOL(sdei_watchdog_callback);
 
 static void sdei_nmi_watchdog_bind(void *data)
 {
--- a/kernel/watchdog.c
+++ b/kernel/watchdog.c
@@ -16,6 +16,7 @@
 #include <linux/cpu.h>
 #include <linux/nmi.h>
 #include <linux/init.h>
+#include <linux/kprobes.h>
 #include <linux/module.h>
 #include <linux/sysctl.h>
 #include <linux/tick.h>
@@ -311,6 +312,7 @@ bool is_hardlockup(void)
 	__this_cpu_write(hrtimer_interrupts_saved, hrint);
 	return false;
 }
+NOKPROBE_SYMBOL(is_hardlockup);
 
 static void watchdog_interrupt_count(void)
 {
--- a/kernel/watchdog_hld.c
+++ b/kernel/watchdog_hld.c
@@ -14,6 +14,7 @@
 
 #include <linux/nmi.h>
 #include <linux/atomic.h>
+#include <linux/kprobes.h>
 #include <linux/module.h>
 #include <linux/sched/debug.h>
 
@@ -180,6 +181,7 @@ static void watchdog_overflow_callback(s
 
 	watchdog_hardlockup_check(regs);
 }
+NOKPROBE_SYMBOL(watchdog_hardlockup_check);
 
 static int hardlockup_detector_event_create(void)
 {
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -967,7 +967,7 @@ config HARDLOCKUP_DETECTOR_PERF
 
 config SDEI_WATCHDOG
 	bool "SDEI NMI Watchdog support"
-	depends on ARM_SDE_INTERFACE
+	depends on ARM_SDE_INTERFACE && !HARDLOCKUP_CHECK_TIMESTAMP
 	select HAVE_HARDLOCKUP_DETECTOR_ARCH
 	select HARDLOCKUP_DETECTOR
 
